File: sdap-python/src/sdap/handshake/protocol.py
"""SDAP 3-message handshake protocol."""

from __future__ import annotations

import json
import os
import uuid
from dataclasses import dataclass, field
from datetime import datetime, timezone
from enum import Enum
from typing import Callable, Optional

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey
from cryptography.hazmat.primitives.asymmetric.x25519 import X25519PrivateKey

from sdap.crypto import (
    derive_session_keys,
    perform_ecdh,
    sign_jws,
    verify_jws,
)
from sdap.identity.did import DIDDocument, validate_did
from sdap.identity.keys import (
    KeyPair,
    jwk_to_public_key,
    multibase_to_public_key,
    public_key_to_jwk,
)

_MAX_CLOCK_SKEW_SECONDS = 60


class HandshakeState(str, Enum):
    INIT_SENT = "INIT_SENT"
    ACCEPT_RECEIVED = "ACCEPT_RECEIVED"
    CONFIRMED = "CONFIRMED"
    FAILED = "FAILED"


@dataclass
class Session:
    session_id: str
    initiator_did: str
    responder_did: str
    encrypt_key: bytes
    mac_key: bytes
    granted_scopes: list[str]
    security_level: str
    expiry: datetime
    # Maps DID -> last received sequence number for messages FROM that DID (for replay detection)
    sequence_counter: dict[str, int] = field(default_factory=dict)
    # Maps DID -> last sent sequence number for messages BY that DID (for outgoing ordering)
    send_counter: dict[str, int] = field(default_factory=dict)

    def is_expired(self) -> bool:
        return datetime.now(tz=timezone.utc) >= self.expiry


def _now_ts() -> int:
    """Return current UTC time as a Unix timestamp integer."""
    return int(datetime.now(tz=timezone.utc).timestamp())


def _check_timestamp(ts: int) -> None:
    """Raise ValueError if timestamp is outside the allowed clock skew."""
    now = _now_ts()
    if abs(now - ts) > _MAX_CLOCK_SKEW_SECONDS:
        raise ValueError(
            f"Timestamp {ts} is outside allowed clock skew ({_MAX_CLOCK_SKEW_SECONDS}s)"
        )


def _get_auth_public_key(did_doc: DIDDocument):
    """Extract the Ed25519 auth public key from a DIDDocument."""
    if not did_doc.authentication:
        raise ValueError(f"DID document {did_doc.id} has no authentication keys")
    auth_key_id = did_doc.authentication[0]
    # auth_key_id may be a full reference like "did:sdap:...#auth-key-1"
    for vm in did_doc.verificationMethod:
        if vm.id == auth_key_id:
            return multibase_to_public_key(vm.publicKeyMultibase, "Ed25519")
    raise ValueError(f"Auth key {auth_key_id} not found in verification methods")


def _get_agreement_public_key(did_doc: DIDDocument):
    """Extract the X25519 key agreement public key from a DIDDocument."""
    if not did_doc.keyAgreement:
        raise ValueError(f"DID document {did_doc.id} has no key agreement keys")
    agree_key_id = did_doc.keyAgreement[0]
    for vm in did_doc.verificationMethod:
        if vm.id == agree_key_id:
            return multibase_to_public_key(vm.publicKeyMultibase, "X25519")
    raise ValueError(f"Agreement key {agree_key_id} not found in verification methods")


def create_handshake_init(
    initiator_did: str,
    target_did: str,
    auth_private_key: Ed25519PrivateKey,
    auth_key_id: str,
    ephemeral_keypair: KeyPair,
    requested_scopes: list[str],
    required_security_level: str = "standard",
) -> tuple[dict, X25519PrivateKey]:
    """Create a handshake INIT message.

    Returns:
        (init_message_dict, ephemeral_private_key)
    """
    if not validate_did(initiator_did):
        raise ValueError(f"Invalid initiator DID: {initiator_did!r}")
    if not validate_did(target_did):
        raise ValueError(f"Invalid target DID: {target_did!r}")

    nonce = os.urandom(32).hex()
    session_id = str(uuid.uuid4())

    payload = {
        "type": "sdap-handshake-init",
        "version": "1.0",
        "sessionId": session_id,
        "initiatorDID": initiator_did,
        "targetDID": target_did,
        "nonce": nonce,
        "timestamp": _now_ts(),
        "ephemeralKey": public_key_to_jwk(ephemeral_keypair.public_key),
        "requestedScopes": requested_scopes,
        "requiredSecurityLevel": required_security_level,
    }
    payload_bytes = json.dumps(payload, separators=(",", ":")).encode("utf-8")
    jws = sign_jws(payload_bytes, auth_private_key, auth_key_id)

    init_msg = {
        "type": "sdap-handshake-init",
        "jws": jws,
    }
    return init_msg, ephemeral_keypair.private_key


def process_handshake_init(
    init_msg: dict,
    responder_did: str,
    responder_auth_key: Ed25519PrivateKey,
    responder_auth_key_id: str,
    responder_ephemeral: KeyPair,
    resolve_did_func: Callable[[str], DIDDocument],
    granted_scopes: list[str] | None = None,
    session_ttl: int = 3600,
) -> tuple[dict, Session]:
    """Validate an INIT message, produce an ACCEPT message, and derive session keys.

    Returns:
        (accept_msg_dict, session)
    """
    jws = init_msg.get("jws")
    if not jws:
        raise ValueError("Missing 'jws' in init message")

    # Decode without verification first to extract initiator DID
    parts = jws.split(".")
    if len(parts) != 3:
        raise ValueError("Invalid JWS format in init message")

    import base64 as _b64
    payload_b64 = parts[1]
    padding = 4 - len(payload_b64) % 4
    if padding != 4:
        payload_b64 += "=" * padding
    payload_bytes = _b64.urlsafe_b64decode(payload_b64)
    payload = json.loads(payload_bytes)

    initiator_did = payload.get("initiatorDID")
    if not initiator_did or not validate_did(initiator_did):
        raise ValueError(f"Invalid initiatorDID in INIT payload: {initiator_did!r}")

    # Resolve initiator DID to get their auth public key
    initiator_doc = resolve_did_func(initiator_did)
    initiator_auth_key = _get_auth_public_key(initiator_doc)

    # Verify signature
    try:
        verify_jws(jws, initiator_auth_key)
    except ValueError as exc:
        raise ValueError(f"INIT message signature verification failed: {exc}") from exc

    # Validate fields
    if payload.get("type") != "sdap-handshake-init":
        raise ValueError("Invalid message type in INIT payload")
    if payload.get("targetDID") != responder_did:
        raise ValueError(
            f"targetDID mismatch: expected {responder_did!r}, got {payload.get('targetDID')!r}"
        )
    _check_timestamp(payload["timestamp"])

    session_id = payload["sessionId"]
    initiator_nonce = payload["nonce"]
    initiator_ephemeral_key = jwk_to_public_key(payload["ephemeralKey"])
    requested_scopes = payload.get("requestedScopes", [])

    # Determine granted scopes (default: grant all requested)
    if granted_scopes is None:
        granted_scopes = requested_scopes

    # ECDH with initiator's ephemeral key
    shared_secret = perform_ecdh(responder_ephemeral.private_key, initiator_ephemeral_key)

    responder_nonce = os.urandom(32).hex()

    # Derive session keys
    encrypt_key, mac_key = derive_session_keys(
        shared_secret,
        bytes.fromhex(initiator_nonce) if len(initiator_nonce) == 64 else initiator_nonce.encode(),
        responder_nonce.encode(),
        session_id,
    )

    expiry = datetime.fromtimestamp(
        _now_ts() + session_ttl, tz=timezone.utc
    )
    session = Session(
        session_id=session_id,
        initiator_did=initiator_did,
        responder_did=responder_did,
        encrypt_key=encrypt_key,
        mac_key=mac_key,
        granted_scopes=granted_scopes,
        security_level=payload.get("requiredSecurityLevel", "standard"),
        expiry=expiry,
        sequence_counter={initiator_did: 0, responder_did: 0},
    )

    # Build accept payload
    accept_payload = {
        "type": "sdap-handshake-accept",
        "version": "1.0",
        "sessionId": session_id,
        "initiatorDID": initiator_did,
        "responderDID": responder_did,
        "initiatorNonce": initiator_nonce,
        "responderNonce": responder_nonce,
        "timestamp": _now_ts(),
        "ephemeralKey": public_key_to_jwk(responder_ephemeral.public_key),
        "grantedScopes": granted_scopes,
        "securityLevel": session.security_level,
        "sessionExpiry": int(expiry.timestamp()),
    }
    accept_payload_bytes = json.dumps(accept_payload, separators=(",", ":")).encode("utf-8")
    accept_jws = sign_jws(accept_payload_bytes, responder_auth_key, responder_auth_key_id)

    accept_msg = {
        "type": "sdap-handshake-accept",
        "jws": accept_jws,
    }
    return accept_msg, session


def create_handshake_confirm(
    accept_msg: dict,
    initiator_did: str,
    initiator_nonce: str,
    auth_private_key: Ed25519PrivateKey,
    auth_key_id: str,
    initiator_ephemeral_private: X25519PrivateKey,
) -> tuple[dict, Session]:
    """Process an ACCEPT message and produce a CONFIRM message.

    Returns:
        (confirm_msg_dict, session)
    """
    jws = accept_msg.get("jws")
    if not jws:
        raise ValueError("Missing 'jws' in accept message")

    # Decode payload without verification to extract responder DID
    parts = jws.split(".")
    if len(parts) != 3:
        raise ValueError("Invalid JWS format in accept message")

    import base64 as _b64
    payload_b64 = parts[1]
    padding = 4 - len(payload_b64) % 4
    if padding != 4:
        payload_b64 += "=" * padding
    payload_bytes = _b64.urlsafe_b64decode(payload_b64)
    payload = json.loads(payload_bytes)

    if payload.get("type") != "sdap-handshake-accept":
        raise ValueError("Invalid message type in ACCEPT payload")

    responder_did = payload.get("responderDID")
    if not responder_did or not validate_did(responder_did):
        raise ValueError(f"Invalid responderDID in ACCEPT payload: {responder_did!r}")

    # Validate the nonce echo
    if payload.get("initiatorDID") != initiator_did:
        raise ValueError("initiatorDID mismatch in ACCEPT message")
    if payload.get("initiatorNonce") != initiator_nonce:
        raise ValueError("Nonce mismatch in ACCEPT message: initiator nonce not echoed correctly")

    _check_timestamp(payload["timestamp"])

    session_id = payload["sessionId"]
    responder_nonce = payload["responderNonce"]
    responder_ephemeral_key = jwk_to_public_key(payload["ephemeralKey"])
    granted_scopes = payload.get("grantedScopes", [])
    security_level = payload.get("securityLevel", "standard")
    session_expiry_ts = payload.get("sessionExpiry")

    if session_expiry_ts:
        expiry = datetime.fromtimestamp(session_expiry_ts, tz=timezone.utc)
    else:
        expiry = datetime.fromtimestamp(_now_ts() + 3600, tz=timezone.utc)

    # ECDH and key derivation
    shared_secret = perform_ecdh(initiator_ephemeral_private, responder_ephemeral_key)
    encrypt_key, mac_key = derive_session_keys(
        shared_secret,
        bytes.fromhex(initiator_nonce) if len(initiator_nonce) == 64 else initiator_nonce.encode(),
        responder_nonce.encode(),
        session_id,
    )

    session = Session(
        session_id=session_id,
        initiator_did=initiator_did,
        responder_did=responder_did,
        encrypt_key=encrypt_key,
        mac_key=mac_key,
        granted_scopes=granted_scopes,
        security_level=security_level,
        expiry=expiry,
        sequence_counter={initiator_did: 0, responder_did: 0},
    )

    # Build confirm payload
    confirm_payload = {
        "type": "sdap-handshake-confirm",
        "version": "1.0",
        "sessionId": session_id,
        "initiatorDID": initiator_did,
        "responderDID": responder_did,
        "responderNonce": responder_nonce,
        "timestamp": _now_ts(),
    }
    confirm_payload_bytes = json.dumps(confirm_payload, separators=(",", ":")).encode("utf-8")
    confirm_jws = sign_jws(confirm_payload_bytes, auth_private_key, auth_key_id)

    confirm_msg = {
        "type": "sdap-handshake-confirm",
        "jws": confirm_jws,
    }
    return confirm_msg, session


def process_handshake_confirm(
    confirm_msg: dict,
    session: Session,
    resolve_did_func: Callable[[str], DIDDocument],
) -> Session:
    """Validate a CONFIRM message and return the confirmed session.

    Returns:
        Session (same object, validated)
    """
    jws = confirm_msg.get("jws")
    if not jws:
        raise ValueError("Missing 'jws' in confirm message")

    # Decode payload
    parts = jws.split(".")
    if len(parts) != 3:
        raise ValueError("Invalid JWS format in confirm message")

    import base64 as _b64
    payload_b64 = parts[1]
    padding = 4 - len(payload_b64) % 4
    if padding != 4:
        payload_b64 += "=" * padding
    payload_bytes = _b64.urlsafe_b64decode(payload_b64)
    payload = json.loads(payload_bytes)

    if payload.get("type") != "sdap-handshake-confirm":
        raise ValueError("Invalid message type in CONFIRM payload")

    if payload.get("sessionId") != session.session_id:
        raise ValueError("Session ID mismatch in CONFIRM message")
    if payload.get("initiatorDID") != session.initiator_did:
        raise ValueError("initiatorDID mismatch in CONFIRM message")
    if payload.get("responderDID") != session.responder_did:
        raise ValueError("responderDID mismatch in CONFIRM message")

    _check_timestamp(payload["timestamp"])

    if session.is_expired():
        raise ValueError("Session has expired before confirmation")

    # Verify signature using initiator's auth key
    initiator_doc = resolve_did_func(session.initiator_did)
    initiator_auth_key = _get_auth_public_key(initiator_doc)

    try:
        verify_jws(jws, initiator_auth_key)
    except ValueError as exc:
        raise ValueError(f"CONFIRM message signature verification failed: {exc}") from exc

    return session

File: sdap-typescript/src/handshake/protocol.ts
/**
 * SDAP 3-message handshake protocol.
 */

import { randomBytes } from "node:crypto";
import { randomUUID } from "node:crypto";
import {
  signJws,
  verifyJws,
  performEcdh,
  deriveSessionKeys,
} from "../crypto/index.js";
import { validateDid, DIDDocument } from "../identity/did.js";
import {
  KeyPair,
  publicKeyToJwk,
  jwkToPublicKey,
  multibaseToPublicKey,
} from "../identity/keys.js";

const MAX_CLOCK_SKEW_SECONDS = 60;

export interface Session {
  sessionId: string;
  initiatorDid: string;
  responderDid: string;
  encryptKey: Uint8Array;
  macKey: Uint8Array;
  grantedScopes: string[];
  securityLevel: string;
  expiry: Date;
  /** Maps DID → last received sequence number (for replay detection) */
  sequenceCounter: Record<string, number>;
  /** Maps DID → last sent sequence number (for outgoing ordering) */
  sendCounter: Record<string, number>;
}

export function isSessionExpired(session: Session): boolean {
  return new Date() >= session.expiry;
}

function nowTs(): number {
  return Math.floor(Date.now() / 1000);
}

function checkTimestamp(ts: number): void {
  const now = nowTs();
  if (Math.abs(now - ts) > MAX_CLOCK_SKEW_SECONDS) {
    throw new Error(
      `Timestamp ${ts} is outside allowed clock skew (${MAX_CLOCK_SKEW_SECONDS}s)`
    );
  }
}

function getAuthPublicKey(didDoc: DIDDocument): Uint8Array {
  if (!didDoc.authentication || didDoc.authentication.length === 0) {
    throw new Error(`DID document ${didDoc.id} has no authentication keys`);
  }
  const authKeyId = didDoc.authentication[0];
  for (const vm of didDoc.verificationMethod) {
    if (vm.id === authKeyId) {
      return multibaseToPublicKey(vm.publicKeyMultibase);
    }
  }
  throw new Error(
    `Auth key ${authKeyId} not found in verification methods`
  );
}

function getAgreementPublicKey(didDoc: DIDDocument): Uint8Array {
  if (!didDoc.keyAgreement || didDoc.keyAgreement.length === 0) {
    throw new Error(`DID document ${didDoc.id} has no key agreement keys`);
  }
  const agreeKeyId = didDoc.keyAgreement[0];
  for (const vm of didDoc.verificationMethod) {
    if (vm.id === agreeKeyId) {
      return multibaseToPublicKey(vm.publicKeyMultibase);
    }
  }
  throw new Error(
    `Agreement key ${agreeKeyId} not found in verification methods`
  );
}

function decodeJwsPayload(jws: string): Record<string, unknown> {
  const parts = jws.split(".");
  if (parts.length !== 3) {
    throw new Error("Invalid JWS format");
  }
  const payloadB64 = parts[1];
  const padding = 4 - (payloadB64.length % 4);
  const padded = padding !== 4 ? payloadB64 + "=".repeat(padding) : payloadB64;
  const b64 = padded.replace(/-/g, "+").replace(/_/g, "/");
  return JSON.parse(Buffer.from(b64, "base64").toString("utf8"));
}

/**
 * Create a handshake INIT message.
 *
 * Returns { initMessage, ephemeralPrivateKey }
 */
export async function createHandshakeInit(params: {
  initiatorDid: string;
  targetDid: string;
  authPrivateKey: Uint8Array;
  authKeyId: string;
  ephemeralKeypair: KeyPair;
  requestedScopes: string[];
  requiredSecurityLevel?: string;
}): Promise<{ initMessage: Record<string, unknown>; ephemeralPrivateKey: Uint8Array }> {
  const {
    initiatorDid,
    targetDid,
    authPrivateKey,
    authKeyId,
    ephemeralKeypair,
    requestedScopes,
    requiredSecurityLevel = "standard",
  } = params;

  if (!validateDid(initiatorDid)) {
    throw new Error(`Invalid initiator DID: ${JSON.stringify(initiatorDid)}`);
  }
  if (!validateDid(targetDid)) {
    throw new Error(`Invalid target DID: ${JSON.stringify(targetDid)}`);
  }

  const nonce = randomBytes(32).toString("hex");
  const sessionId = randomUUID();

  const payload = {
    type: "sdap-handshake-init",
    version: "1.0",
    sessionId,
    initiatorDID: initiatorDid,
    targetDID: targetDid,
    nonce,
    timestamp: nowTs(),
    ephemeralKey: publicKeyToJwk(ephemeralKeypair.publicKey),
    requestedScopes,
    requiredSecurityLevel,
  };

  const payloadBytes = new TextEncoder().encode(JSON.stringify(payload));
  const jws = await signJws(payloadBytes, authPrivateKey, authKeyId);

  const initMessage = {
    type: "sdap-handshake-init",
    jws,
  };

  return { initMessage, ephemeralPrivateKey: ephemeralKeypair.privateKey };
}

/**
 * Validate an INIT message, produce an ACCEPT message, and derive session keys.
 *
 * Returns { acceptMessage, session }
 */
export async function processHandshakeInit(params: {
  initMsg: Record<string, unknown>;
  responderDid: string;
  responderAuthKey: Uint8Array;
  responderAuthKeyId: string;
  responderEphemeral: KeyPair;
  resolveDidFn: (did: string) => DIDDocument;
  grantedScopes?: string[];
  sessionTtl?: number;
}): Promise<{ acceptMessage: Record<string, unknown>; session: Session }> {
  const {
    initMsg,
    responderDid,
    responderAuthKey,
    responderAuthKeyId,
    responderEphemeral,
    resolveDidFn,
    grantedScopes: grantedScopesOverride,
    sessionTtl = 3600,
  } = params;

  const jws = initMsg["jws"] as string | undefined;
  if (!jws) {
    throw new Error("Missing 'jws' in init message");
  }

  // Decode without verification first to extract initiator DID
  const payload = decodeJwsPayload(jws);

  const initiatorDid = payload["initiatorDID"] as string | undefined;
  if (!initiatorDid || !validateDid(initiatorDid)) {
    throw new Error(`Invalid initiatorDID in INIT payload: ${JSON.stringify(initiatorDid)}`);
  }

  // Resolve initiator DID to get auth public key
  const initiatorDoc = resolveDidFn(initiatorDid);
  const initiatorAuthKey = getAuthPublicKey(initiatorDoc);

  // Verify signature
  try {
    await verifyJws(jws, initiatorAuthKey);
  } catch (err) {
    throw new Error(`INIT message signature verification failed: ${err}`);
  }

  if (payload["type"] !== "sdap-handshake-init") {
    throw new Error("Invalid message type in INIT payload");
  }
  if (payload["targetDID"] !== responderDid) {
    throw new Error(
      `targetDID mismatch: expected ${JSON.stringify(responderDid)}, got ${JSON.stringify(payload["targetDID"])}`
    );
  }
  checkTimestamp(payload["timestamp"] as number);

  const sessionId = payload["sessionId"] as string;
  const initiatorNonce = payload["nonce"] as string;
  const initiatorEphemeralKey = jwkToPublicKey(
    payload["ephemeralKey"] as JsonWebKey
  );
  const requestedScopes = (payload["requestedScopes"] as string[]) ?? [];

  const grantedScopes = grantedScopesOverride ?? requestedScopes;

  // ECDH with initiator's ephemeral key
  const sharedSecret = performEcdh(
    responderEphemeral.privateKey,
    initiatorEphemeralKey
  );

  const responderNonce = randomBytes(32).toString("hex");

  // nonce_a: initiator nonce (hex string → bytes if 64 hex chars, else UTF-8)
  const nonceABytes =
    initiatorNonce.length === 64
      ? Buffer.from(initiatorNonce, "hex")
      : Buffer.from(initiatorNonce, "utf8");
  const nonceBBytes = Buffer.from(responderNonce, "utf8");

  const { encryptKey, macKey } = deriveSessionKeys(
    sharedSecret,
    nonceABytes,
    nonceBBytes,
    sessionId
  );

  const expiryTs = nowTs() + sessionTtl;
  const expiry = new Date(expiryTs * 1000);

  const securityLevel =
    (payload["requiredSecurityLevel"] as string) ?? "standard";

  const session: Session = {
    sessionId,
    initiatorDid,
    responderDid,
    encryptKey,
    macKey,
    grantedScopes,
    securityLevel,
    expiry,
    sequenceCounter: { [initiatorDid]: 0, [responderDid]: 0 },
    sendCounter: {},
  };

  // Build accept payload
  const acceptPayload = {
    type: "sdap-handshake-accept",
    version: "1.0",
    sessionId,
    initiatorDID: initiatorDid,
    responderDID: responderDid,
    initiatorNonce,
    responderNonce,
    timestamp: nowTs(),
    ephemeralKey: publicKeyToJwk(responderEphemeral.publicKey),
    grantedScopes,
    securityLevel,
    sessionExpiry: expiryTs,
  };

  const acceptPayloadBytes = new TextEncoder().encode(
    JSON.stringify(acceptPayload)
  );
  const acceptJws = await signJws(
    acceptPayloadBytes,
    responderAuthKey,
    responderAuthKeyId
  );

  const acceptMessage = {
    type: "sdap-handshake-accept",
    jws: acceptJws,
  };

  return { acceptMessage, session };
}

/**
 * Process an ACCEPT message and produce a CONFIRM message.
 *
 * Returns { confirmMessage, session }
 */
export async function createHandshakeConfirm(params: {
  acceptMsg: Record<string, unknown>;
  initiatorDid: string;
  initiatorNonce: string;
  authPrivateKey: Uint8Array;
  authKeyId: string;
  initiatorEphemeralPrivate: Uint8Array;
}): Promise<{ confirmMessage: Record<string, unknown>; session: Session }> {
  const {
    acceptMsg,
    initiatorDid,
    initiatorNonce,
    authPrivateKey,
    authKeyId,
    initiatorEphemeralPrivate,
  } = params;

  const jws = acceptMsg["jws"] as string | undefined;
  if (!jws) {
    throw new Error("Missing 'jws' in accept message");
  }

  const payload = decodeJwsPayload(jws);

  if (payload["type"] !== "sdap-handshake-accept") {
    throw new Error("Invalid message type in ACCEPT payload");
  }

  const responderDid = payload["responderDID"] as string | undefined;
  if (!responderDid || !validateDid(responderDid)) {
    throw new Error(
      `Invalid responderDID in ACCEPT payload: ${JSON.stringify(responderDid)}`
    );
  }

  if (payload["initiatorDID"] !== initiatorDid) {
    throw new Error("initiatorDID mismatch in ACCEPT message");
  }
  if (payload["initiatorNonce"] !== initiatorNonce) {
    throw new Error(
      "Nonce mismatch in ACCEPT message: initiator nonce not echoed correctly"
    );
  }

  checkTimestamp(payload["timestamp"] as number);

  const sessionId = payload["sessionId"] as string;
  const responderNonce = payload["responderNonce"] as string;
  const responderEphemeralKey = jwkToPublicKey(
    payload["ephemeralKey"] as JsonWebKey
  );
  const grantedScopes = (payload["grantedScopes"] as string[]) ?? [];
  const securityLevel = (payload["securityLevel"] as string) ?? "standard";
  const sessionExpiryTs = payload["sessionExpiry"] as number | undefined;

  const expiry = sessionExpiryTs
    ? new Date(sessionExpiryTs * 1000)
    : new Date((nowTs() + 3600) * 1000);

  // ECDH and key derivation
  const sharedSecret = performEcdh(
    initiatorEphemeralPrivate,
    responderEphemeralKey
  );

  const nonceABytes =
    initiatorNonce.length === 64
      ? Buffer.from(initiatorNonce, "hex")
      : Buffer.from(initiatorNonce, "utf8");
  const nonceBBytes = Buffer.from(responderNonce, "utf8");

  const { encryptKey, macKey } = deriveSessionKeys(
    sharedSecret,
    nonceABytes,
    nonceBBytes,
    sessionId
  );

  const session: Session = {
    sessionId,
    initiatorDid,
    responderDid,
    encryptKey,
    macKey,
    grantedScopes,
    securityLevel,
    expiry,
    sequenceCounter: { [initiatorDid]: 0, [responderDid]: 0 },
    sendCounter: {},
  };

  // Build confirm payload
  const confirmPayload = {
    type: "sdap-handshake-confirm",
    version: "1.0",
    sessionId,
    initiatorDID: initiatorDid,
    responderDID: responderDid,
    responderNonce,
    timestamp: nowTs(),
  };

  const confirmPayloadBytes = new TextEncoder().encode(
    JSON.stringify(confirmPayload)
  );
  const confirmJws = await signJws(
    confirmPayloadBytes,
    authPrivateKey,
    authKeyId
  );

  const confirmMessage = {
    type: "sdap-handshake-confirm",
    jws: confirmJws,
  };

  return { confirmMessage, session };
}

/**
 * Validate a CONFIRM message and return the confirmed session.
 */
export async function processHandshakeConfirm(params: {
  confirmMsg: Record<string, unknown>;
  session: Session;
  resolveDidFn: (did: string) => DIDDocument;
}): Promise<Session> {
  const { confirmMsg, session, resolveDidFn } = params;

  const jws = confirmMsg["jws"] as string | undefined;
  if (!jws) {
    throw new Error("Missing 'jws' in confirm message");
  }

  const payload = decodeJwsPayload(jws);

  if (payload["type"] !== "sdap-handshake-confirm") {
    throw new Error("Invalid message type in CONFIRM payload");
  }

  if (payload["sessionId"] !== session.sessionId) {
    throw new Error("Session ID mismatch in CONFIRM message");
  }
  if (payload["initiatorDID"] !== session.initiatorDid) {
    throw new Error("initiatorDID mismatch in CONFIRM message");
  }
  if (payload["responderDID"] !== session.responderDid) {
    throw new Error("responderDID mismatch in CONFIRM message");
  }

  checkTimestamp(payload["timestamp"] as number);

  if (isSessionExpired(session)) {
    throw new Error("Session has expired before confirmation");
  }

  // Verify signature using initiator's auth key
  const initiatorDoc = resolveDidFn(session.initiatorDid);
  const initiatorAuthKey = getAuthPublicKey(initiatorDoc);

  try {
    await verifyJws(jws, initiatorAuthKey);
  } catch (err) {
    throw new Error(`CONFIRM message signature verification failed: ${err}`);
  }

  return session;
}
